### ISSUE TITLE
Test Failure Overlay in Wrong Location

### ISSUE THREAD

**Issue by AshtonKem** [clojure-emacs/cider#1532]
I have some tests that depend on assertion helpers written in another file. I've noticed since upgrading to Cider 0.11.0-SNAPSHOT that the overlays are now being displayed in the wrong place in the test buffer. 

More specifically, let's say I have test_helpers.clj, which has an assertion helper function on line 316. In my test namespace, any test that uses this helper will have their failure placed on line 316 of the _test_ buffer, regardless of where in the buffer the failing test actually is. 

`CIDER 0.11.0snapshot (package: 20160125.741) (Java 1.8.0_66, Clojure 1.7.0, nREPL 0.2.12)`

**Comment by bbatsov** [comment]
Some mini repro project would help a lot in debugging this.

**Comment by AshtonKem** [comment]
I will try to get to that this weekend. 

--
Ashton

**Comment by AshtonKem** [comment]
I'm trying to reproduce this in a cleaner repo, and I'm getting the following error instead:

``` 
error in process filter: Not a nREPL dict object: dict
```

I am unsure how to diagnose if this is the same issue or unrelated. How do I do that?

**Comment by Malabarba** [comment]
It's a new issue. I get it too. Probably due to the recent addition to the test commands.

**Comment by Malabarba** [comment]
Forgot to ping @jeffvalk.

**Comment by jeffvalk** [comment]
The `Not a nREPL dict object` error sounds a bit like the `cider` and `cider-nrepl` versions were out of sync. If you're still seeing this with the latest of each, can you post a repro case?

**Comment by Malabarba** [comment]
I'm sorry, I think you're right @jeffvalk 
I restarted Emacs and don't get any errors now.

**Comment by jeffvalk** [comment]
@Malabarba Ok, good.

Still waiting on an example for the original issue.

**Comment by AshtonKem** [comment]
Restarting emacs did not resolve this issue for me, and it's now preventing me from creating an example case for this bug (no overlay appears after the above error message). I'll open a new issue for that bug.

**Comment by Malabarba** [comment]
@ashtonkem It's good to double check that you're updated on both sides of cider.

**Comment by AshtonKem** [comment]
I have created a test repo here: https://github.com/AshtonKem/cider-example

Simple open core_test.clj and test either the entire NS or just the individual test, doesn't make a difference in the result.

**Comment by bbatsov** [comment]
@jeffvalk can you take a look at this?

**Comment by jeffvalk** [comment]
@bbatsov Will do. It may be later in the week though.

**Comment by bbatsov** [comment]
No problem.

**Comment by bbatsov** [comment]
P.S. If anybody else would like to investigate this - be my guest.

**Comment by xiongtx** [comment]
I don't think there's an easy fix for this.

In `cider-test.el`, we [purposefully ignore the file](https://github.com/clojure-emacs/cider/blob/1ee97c4c4bbb9f7c8b6dffde0112cbb8338870dd/cider-test.el#L517-L519), presumably b/c we assume that all assertions are in the current buffer.

If this is not the case, there are a few possibilities:

- There is a file and it's not the buffer's file
  - _Should_ we visit the file and highlight the issue there?
    - I [wouldn't expect](http://wiki.c2.com/?PrincipleOfLeastAstonishment) `cider-test` to pop up new file buffers.
  - _Can_ we visit the file? 
    - `clojure.test`_ [only gives the file name](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/test.clj#L348), not url.
- There is no file (`buffer-file-name` is `nil`).
  - This is perfectly valid; the user can create a buffer and write some test forms w/out saving it to a file

**Comment by bbatsov** [comment]
> More specifically, let's say I have test_helpers.clj, which has an assertion helper function on line 316. In my test namespace, any test that uses this helper will have their failure placed on line 316 of the test buffer, regardless of where in the buffer the failing test actually is.

Well, I guess for such cases we can at least check that the filename matches the one of the current buffer. 

But I agree that's a tricky issue to fix - it will likely be strange to get a new file opened if the helpers are defined there.

**Comment by bbatsov** [comment]
Seems we're out of ideas on how to proceed here.
